Support for syncing units from libraries to courses

navinkarkera opened the pull request:
## Description

Sync unit containers from library to courses.

* **Related to:** https://github.com/openedx/frontend-app-authoring/issues/1711
* `Private-ref`: [FAL-4077](https://tasks.opencraft.com/browse/FAL-4077)

## Test instructions:

Run the new integration tests, which should cover most things:

```
DJANGO_SETTINGS_MODULE=cms.envs.tutor.test pytest cms/djangoapps/contentstore/rest_api/v2/views/tests/test_downstream_sync_integration.py
```


## Manual Test instructions:

Use curl or any other tool to post request, NOTE: below example does not contain `csrftoken`, `studio_session_id` etc. which you can add. TIP: use developer console to copy any request as curl request in browser.
```sh
curl 'http://studio.local.openedx.io:8001/xblock/' \
  -H 'Accept: application/json, text/plain, */*' \
  -H 'Accept-Language: en-US,en;q=0.6' \
  -H 'Connection: keep-alive' \
  -H 'Content-Type: application/json' \
  --data-raw '{"category":"vertical","parent_locator":"block-v1:OpenedX+DemoX+DemoCourse+type@sequential+block@727232bd02b54af08953c66eb2246d00","library_content_key":"lct:SampleTaxonomyOrg1:lib1:unit:images-57b324"}' \
  --insecure
```

navinkarkera committed:
feat: add support for lct in library_context

navinkarkera committed:
feat: library unit sync

openedx-webhooks commented:
Thanks for the pull request, @navinkarkera!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

navinkarkera committed:
fixup! feat: library unit sync

navinkarkera committed:
fixup! feat: library unit sync

navinkarkera committed:
fixup! feat: library unit sync

navinkarkera committed:
feat: unit sync using emulated library container usage locator key

navinkarkera committed:
Revert "feat: add support for lct in library_context"

This reverts commit 72ba6ef2ae6ffb5c83670d522b75c200810fca17.

navinkarkera committed:
Revert "feat: unit sync using emulated library container usage locator key"

This reverts commit e8a8d58b10d182aa552cacfc2641db7bc791f619.

navinkarkera committed:
feat: separate managers for component and container upstream sync

navinkarkera committed:
fixup! feat: separate managers for component and container upstream sync

navinkarkera committed:
feat: remove upstream links from unit child components on sync

navinkarkera committed:
refactor!: rename publishable entity links table and update upstream_block field

navinkarkera committed:
feat: create component link only for component xblocks

bradenmacdonald commented on cms/lib/xblock/container_upstream_sync.py:
```diff
@@ -0,0 +1,202 @@
+                if field.scope in [Scope.settings, Scope.content]
+            )
+            for block in [self.upstream, self.downstream]
+        ])
```
For Teak, containers probably won't have any fields other than `display_name`, so that's likely the only one necessary to sync. Also container fields don't have scopes.

If we manage to implement https://github.com/openedx/frontend-app-authoring/issues/1640 soon, we may have two actual fields on units, but we can add the field syncing then.

bradenmacdonald commented on openedx/core/djangoapps/xblock/api.py:
```diff
@@ -70,7 +71,7 @@ def get_runtime(user: UserType):
 
 def load_block(
-    usage_key: UsageKeyV2,
+    usage_key: UsageKeyV2 | LibraryContainerLocator,
```
I'm assuming this is old, and we don't need to support containers in `load_block` anymore ?

We can remove `get_container_block` and `library_container_xml` and a bunch of other changes.

navinkarkera replied:
@bradenmacdonald We still need to load unit xblock (in course they are xblocks) so it was easier to just update this method.

bradenmacdonald replied:
I'm not sure I follow. In the course, the Unit containers will have been converted to `vertical` XBlocks, and can be loaded via modulestore APIs like `get_item()`

ormsbee commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -124,22 +124,22 @@ def upstream_version(self) -> int | None:
+        if hasattr(self.upstream_block.publishable_entity, 'published'):
+            if hasattr(self.upstream_block.publishable_entity.published, 'version'):
+                if hasattr(self.upstream_block.publishable_entity.published.version, 'version_num'):
+                    version_num = self.upstream_block.publishable_entity.published.version.version_num
```
FWIW, we should be able to simplify this down to use `openedx_learning.api.authoring.get_published_version`, like:

```python
published_version = get_published_version(self.upstream_block.id)
return published_version.version_num if published_version else None
```

navinkarkera committed:
feat: container link model

navinkarkera committed:
chore: fix lint issues

navinkarkera committed:
temp: point opaque keys to dev branch

navinkarkera committed:
fix: mypy errors

navinkarkera committed:
fix: import lint errors

navinkarkera committed:
fix: failing tests

navinkarkera committed:
fix: doc lint issues

navinkarkera committed:
fix: lint issues

navinkarkera committed:
fix: failing tests

navinkarkera committed:
Revert "fix: failing tests"

This reverts commit fe316a18eca75ec6815523ae27e5c7130e6fe006.

navinkarkera committed:
feat: update downstream api views

navinkarkera committed:
feat: delete extra components in container on sync (not working)

bradenmacdonald committed:
chore: update with latest master

bradenmacdonald committed:
chore: update with latest master

bradenmacdonald committed:
revert: changes to xblock runtime

bradenmacdonald committed:
chore: cleanups

bradenmacdonald committed:
revert: changes to upstream_sync.py

bradenmacdonald committed:
feat: WIP toward syncing

bradenmacdonald committed:
fix: duplicate definitions of LibraryXBlockMetadata

bradenmacdonald committed:
fix: type issues

bradenmacdonald committed:
test: fix tests

bradenmacdonald committed:
fix: duplicate definitions of LibraryXBlockMetadata (fixup)

bradenmacdonald committed:
chore: quality issues

bradenmacdonald committed:
test: update tests

bradenmacdonald committed:
test: add a new integration test suite for syncing

bradenmacdonald committed:
test: continue building out integration tests

bradenmacdonald committed:
feat: partially implement container+child syncing

bradenmacdonald committed:
test: continue building out integration tests

bradenmacdonald committed:
test: continue building out integration tests

bradenmacdonald committed:
fix: blockserializer wasn't always serializing all HTML block fields

bradenmacdonald committed:
test: continue building out integration tests

bradenmacdonald committed:
chore: quality fix

bradenmacdonald committed:
chore: updated query count in test, from serializer/dataclass change

bradenmacdonald commented:
Handing this over to @navinkarkera now. @pomegranited @kdmccormick @ormsbee can you help get this merged as soon as it's ready? (as well as any PRs we need that depend on these APIs)

I undid a bunch of earlier changes, mostly because I'm trying to keep the changes to the (working) sync code for Containers as minimal as possible, although I still ended up doing a lot of refactoring.

Note that I think the "Manual Test instructions" are out of date; I have mostly been working on the integration tests in `test_downstream_sync_integration.py`. But specifically, the API request needs to have `category: vertical` even though it's a unit, and I don't think it needs `"type":"library_v2"`. See comments in `test_downstream_sync_integration.py` for details.

The integration tests are based purely on the REST APIs (with one tiny exception), so anything that you see working in the tests should be working/workable from the UI too. I recommend reading through [the integration tests](https://github.com/open-craft/edx-platform/blob/ecdeafd6a00c3e70d1f49b223c15b5882752fd19/cms/djangoapps/contentstore/rest_api/v2/views/tests/test_downstream_sync_integration.py#L120) to get a sense of how this works and what data/operations are available.

TODO before we can merge:
- [x] get the build green
- [x] make sure there's no regressions to `ComponentLink` tracking and `recreate_upstream_links`
- [x] manually test to ensure that syncing regular (non-unit) components from libraries to courses is still working fine (no regressions)

TODO if there's time, otherwise as a fast follow:
- [ ] Finish implementing tracking of container links via `ContainerLink` - I'm not sure of the status
- [x] Finish up any necessary changes to `recreate_upstream_links` to support containers - I'm not sure of the status
- [ ] fix the upstream link status so that if a unit is not modified upstream but it contains a modified component, it shows `ready_to_sync: True`. See [here](https://github.com/open-craft/edx-platform/blob/ecdeafd6a00c3e70d1f49b223c15b5882752fd19/cms/djangoapps/contentstore/rest_api/v2/views/tests/test_downstream_sync_integration.py#L296-L298).

TODO ASAP but I don't want to block merge on it:
- [x] improve the sophistication of the syncing algorithm to avoid losing student state if an upstream (or downstream) block is added or deleted from the beginning/middle of the child component list. Currently the "child sync" algorithm is very very very simple and will just delete any downstream blocks that don't directly correspond to the block in the same position upstream. We don't allow editing downstream units, but we do allow adding components to upstream units so this could be an issue.

bradenmacdonald commented:
```
cms.djangoapps.contentstore.rest_api.v2.views.tests.test_downstream_sync_integration:14: imported from non-public API of openedx.core.djangoapps.content_libraries:
    from openedx.core.djangoapps.content_libraries.tests.base import ContentLibrariesRestApiTest
```
to fix this error, it will be necessary to move the base class somewhere else, or add an exception to the import linter in `setup.cfg`.

Based on my recent PR https://github.com/openedx/edx-platform/pull/36581 , it's perfectly fine for the upstream_sync code to import content libraries code.

navinkarkera committed:
fix: import lint issue

navinkarkera committed:
Merge branch 'master' into navin/fal-4077/link-unit

navinkarkera committed:
fix: missing field in library xblock metadata

navinkarkera committed:
test: update upstream sync error message in tests

pomegranited commented on cms/lib/xblock/upstream_sync.py:
```diff
@@ -138,201 +144,81 @@ def try_get_for_block(cls, downstream: XBlock) -> t.Self:
+                container_meta = lib_api.get_container(upstream_key)
+            except lib_api.ContentLibraryContainerNotFound as exc:
+                raise BadUpstream(_("Linked upstream library container was not found in the system")) from exc
+            expected_downstream_block_type = container_meta.container_type.olx_tag
+            version_available = container_meta.published_version_num
+
+        if downstream_type != expected_downstream_block_type:
```
I'm getting an error here when I try to add a library Unit to a course, because:

```
downstream_type == 'unit'
expected_downstream_block_type == 'vertical'
```

Is the downstream block being created with the wrong block type? Or should this be this instead (so that `expected_downstream_block_type  == 'unit'` too)?

```suggestion
            expected_downstream_block_type = container_meta.container_type.value
            version_available = container_meta.published_version_num

        if downstream_type != expected_downstream_block_type:
```

navinkarkera replied:
@pomegranited You need to specify `category: "vertical"` in the API call. See Braden's comment above:

> Note that I think the "Manual Test instructions" are out of date; I have mostly been working on the integration tests in test_downstream_sync_integration.py. But specifically, the API request needs to have category: vertical even though it's a unit, and I don't think it needs "type":"library_v2". See comments in test_downstream_sync_integration.py for details.

bradenmacdonald replied:
Yep - I mentioned this in the test case as well:

```
# The API consumer needs to specify "vertical" here, even though upstream is "unit".
# In the future we could create a nicer REST API endpoint for this that's not part of
# the messy '/xblock/' API and which auto-detects the types based on the upstream_key.
```

@navinkarkera had some code to modify the `/xblock/` endpoint to override this automatically, but I'm a little cautious about changing that existing API endpoint too much (since `unit` _is_ also a valid XBlock some people might want to create - we used it for LabXchange). So I'd much rather we make a new API in the future for this thing, like `/api/contentstore/v2/downstreams/clone/` or something like that, and the new endpoint can auto-set `vertical` for you based on the source/upstream container type, using the nice new code in https://github.com/openedx/edx-platform/pull/36580

navinkarkera committed:
revert: delete_item api changes

pomegranited commented on openedx/core/djangoapps/content_libraries/api/containers.py:
```diff
@@ -372,7 +373,7 @@ def restore_container(container_key: LibraryContainerLocator) -> None:
 def get_container_children(
     container_key: LibraryContainerLocator,
     published=False,
```
This is unrelated to this PR, but a bad conflict merge broke [this line](https://github.com/openedx/edx-platform/pull/36553/files#diff-5a248b18ae17c86d454776cb48e0372a8cbdc9610b3886f175a116f9b280e681R346) above. Could you fix this here?

```diff
@@ -343,7 +343,7 @@ def restore_container(container_key: LibraryContainerLocator) -> None:

     LIBRARY_CONTAINER_CREATED.send_event(
         library_container=LibraryContainerData(
-            container_key=str(container_key),
+            container_key=container_key,
         )
     )
```

navinkarkera replied:
Sure.

navinkarkera committed:
feat: handle reorder, addition and deletion of components in sync

Updates children components of unit in course based on upstream unit,
deletes removed component, adds new ones and updates order as per
upstream.

navinkarkera committed:
chore: fix unrelated conflict

navinkarkera committed:
fix: lint issues

pomegranited reviewed (approved):
:+1: Amazing work here @navinkarkera and @bradenmacdonald !

- [x] I tested this branch along with changes from https://github.com/open-craft/edx-platform/pull/774 + https://github.com/open-craft/edx-platform/pull/773 while following the test instructions for https://github.com/openedx/frontend-app-authoring/pull/1829 + https://github.com/openedx/frontend-app-authoring/pull/1850 + https://github.com/openedx/frontend-app-authoring/pull/1833
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] Adds and updates documentation
- [x] <strike>User-facing strings are extracted for translation</strike> N/A

navinkarkera commented:
> TODO before we can merge:
> 
> * [x]  get the build green
> * [x]  make sure there's no regressions to `ComponentLink` tracking and `recreate_upstream_links`
> * [x]  manually test to ensure that syncing regular (non-unit) components from libraries to courses is still working fine (no regressions)
> 
> 
> TODO if there's time, otherwise as a fast follow:
> 
> * [ ]  Finish implementing tracking of container links via `ContainerLink` - I'm not sure of the status
> * [x]  Finish up any necessary changes to `recreate_upstream_links` to support containers - I'm not sure of the status
> * [ ]  fix the upstream link status so that if a unit is not modified upstream but it contains a modified component, it shows `ready_to_sync: True`. See [here](https://github.com/open-craft/edx-platform/blob/ecdeafd6a00c3e70d1f49b223c15b5882752fd19/cms/djangoapps/contentstore/rest_api/v2/views/tests/test_downstream_sync_integration.py#L296-L298).
> 
> TODO ASAP but I don't want to block merge on it:
> 
> * [x]  improve the sophistication of the syncing algorithm to avoid losing student state if an upstream (or downstream) block is added or deleted from the beginning/middle of the child component list. Currently the "child sync" algorithm is very very very simple and will just delete any downstream blocks that don't directly correspond to the block in the same position upstream. We don't allow editing downstream units, but we do allow adding components to upstream units so this could be an issue.

@bradenmacdonald @pomegranited @ormsbee @kdmccormick This is ready for review now. 

> * [ ]  Finish implementing tracking of container links via `ContainerLink` - I'm not sure of the status

This is mostly done, we just need to decide if we need [list and summary view](https://github.com/open-craft/edx-platform/blob/45e9a34fe31950f5a233732c03ec7508717acee0/cms/djangoapps/contentstore/rest_api/v2/views/downstreams.py#L161-L223) api for container links as well or if we want to merge them into a single list view.

> * [ ]  fix the upstream link status so that if a unit is not modified upstream but it contains a modified component, it shows `ready_to_sync: True`. See [here](https://github.com/open-craft/edx-platform/blob/ecdeafd6a00c3e70d1f49b223c15b5882752fd19/cms/djangoapps/contentstore/rest_api/v2/views/tests/test_downstream_sync_integration.py#L296-L298).

I don't want to block the PR for this one, but while this is being reviewed and merged we can work on this.

ChrisChV committed:
feat: return unit upstreamInfo and disallow edits to units in courses that are sourced from a library (#773)

* feat: Add upstream_info to unit
* feat: disallow edits to units in courses that are sourced from a library (#774)

---------

Co-authored-by: Jillian Vogel <jill@opencraft.com>
Co-authored-by: Rômulo Penido <romulo.penido@gmail.com>

bradenmacdonald commented:
Awesome work @navinkarkera 👏🏻 :)

> This is mostly done, we just need to decide if we need [list and summary view](https://github.com/open-craft/edx-platform/blob/45e9a34fe31950f5a233732c03ec7508717acee0/cms/djangoapps/contentstore/rest_api/v2/views/downstreams.py#L161-L223) api for container links as well or if we want to merge them into a single list view.

Can you elaborate on that? And I assume it can happen in a follow-up PR?

navinkarkera commented:
@bradenmacdonald yes, this can be done in follow up PR. If we want to include units in the course libraries page, we'll need to implement list and summary api for container links same as component links.

navinkarkera commented:
@bradenmacdonald yes, this can be done in follow up PR. If we want to include units in the course libraries page, we'll need to implement list and summary api for container links same as component links.

ormsbee commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -115,31 +103,107 @@ class PublishableEntityLink(models.Model):
+        abstract = True
 
     @property
     def upstream_version(self) -> int | None:
```
```suggestion
    def upstream_version_num(self) -> int | None:
```

When we say "version" in openedx-learning stuff, we almost always mean the actual version model.

bradenmacdonald replied:
It looks like the OLX field for this [has been called just `upstream_version` already in Sumac](https://github.com/openedx/edx-platform/blob/a63a6a6fbaec6bf1c234b9b26b66c7b50c00e336/cms/lib/xblock/upstream_sync.py#L396-L404) so it's probably too late to change that, but I can change it on this `EntityLinkBase`/`ComponentLink`/`ContainerLink` model.

bradenmacdonald replied:
Updated, but only in this Python/Django model, and not in the XBlock mixin nor REST API. 959eadd76d4733d3d14cba2fe9e1d79e473b1ac5

ormsbee commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -80,26 +82,12 @@ class Meta:
-        )
-    )
     # Search by library/upstream context key
     upstream_context_key = key_field(
```
Should there also be an fkey to the `LearningPackage` this comes from?

ormsbee replied:
... I guess that's implicit in the upstream block fkeys...

ormsbee commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -115,31 +103,107 @@ class PublishableEntityLink(models.Model):
+    ) -> QuerySet["EntityLinkBase"]:
+        """
+        Get all links along with sync flag, upstream context title and version, with optional filtering.
+        """
```
Nit: I think it would be helpful to add info in the docstring to as to how this is intended to handle deleted entries.

bradenmacdonald replied:
This code wasn't added in the PR, and I'm not exactly sure off-hand, but I agree we should expand the docstring. Maybe @navinkarkera can do that as a follow-up, if he doesn't see this now.

ormsbee commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -115,31 +103,107 @@ class PublishableEntityLink(models.Model):
+        return self.upstream_block.publishable_entity.learning_package.title
+
+    @classmethod
+    def filter_links(
```
Non-blocking: This is something that seems like it should go into a custom Manager instead of a classmethod on the base class.

ormsbee commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -115,31 +103,107 @@ class PublishableEntityLink(models.Model):
+class ComponentLink(EntityLinkBase):
+    """
+    This represents link between any two publishable entities or link between publishable entity and a course
+    xblock. It helps in tracking relationship between xblocks imported from libraries and used in different courses.
```
```suggestion
    XBlock. It helps in tracking relationship between XBlocks imported from libraries and used in different courses.
```

bradenmacdonald committed:
docs: capitalization of XBlock

Co-authored-by: David Ormsbee <dave@axim.org>

bradenmacdonald committed:
refactor: (minor) change python property name to reflect type better

ormsbee commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -197,70 +261,90 @@ def update_or_create(
+            'version_declined': version_declined,
+        }
+        if upstream_container:
+            new_values.update(
+                {
+                    'upstream_container': upstream_container,
+                }
+            )
```
Just for my understanding, this is equivalent to:

```python
new_values['upstream_container'] = upstream_container
```

?

bradenmacdonald replied:
Yeah, as far as I can see. I suspect there were multiple fields being added there before some refactor.

bradenmacdonald replied:
Actually this code is also unchanged from what's already in `master`

navinkarkera replied:
@bradenmacdonald Yes, I think while writing this, I had multiple fields but were removed while refactoring. I'll update it in a follow up PR.

ormsbee commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -197,70 +261,90 @@ def update_or_create(
+            # modified downstream version.
+            # pylint: disable=line-too-long
+            # 1. https://github.com/open-craft/frontend-app-course-authoring/blob/0443d88824095f6f65a3a64b77244af590d4edff/src/course-libraries/ReviewTabContent.tsx#L222-L233
+            has_changes = True  # change to false once above condition is met.
+            for key, value in new_values.items():
+                prev = getattr(link, key)
+                # None != None is True, so we need to check for it specially
+                if prev != value and ~(prev is None and value is None):
+                    has_changes = True
+                    setattr(link, key, value)
```
Just quick feedback that this was a pretty confusing to me.

ormsbee commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -197,70 +261,90 @@ def update_or_create(
+            # modified downstream version.
+            # pylint: disable=line-too-long
+            # 1. https://github.com/open-craft/frontend-app-course-authoring/blob/0443d88824095f6f65a3a64b77244af590d4edff/src/course-libraries/ReviewTabContent.tsx#L222-L233
+            has_changes = True  # change to false once above condition is met.
```
Where would this ever get set to False?

bradenmacdonald replied:
Note for reference: see below for explanation - https://github.com/openedx/edx-platform/pull/36553#issuecomment-2828507128

ormsbee commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -197,70 +261,90 @@ def update_or_create(
+            # pylint: disable=line-too-long
+            # 1. https://github.com/open-craft/frontend-app-course-authoring/blob/0443d88824095f6f65a3a64b77244af590d4edff/src/course-libraries/ReviewTabContent.tsx#L222-L233
+            has_changes = True  # change to false once above condition is met.
+            for key, value in new_values.items():
```
```suggestion
            for key, new_value in new_values.items():
```

ormsbee commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -197,70 +261,90 @@ def update_or_create(
+            # 1. https://github.com/open-craft/frontend-app-course-authoring/blob/0443d88824095f6f65a3a64b77244af590d4edff/src/course-libraries/ReviewTabContent.tsx#L222-L233
+            has_changes = True  # change to false once above condition is met.
+            for key, value in new_values.items():
+                prev = getattr(link, key)
```
```suggestion
                prev_value = getattr(link, key)
```

ormsbee commented on cms/djangoapps/contentstore/models.py:
```diff
@@ -197,70 +261,90 @@ def update_or_create(
+            has_changes = True  # change to false once above condition is met.
+            for key, value in new_values.items():
+                prev = getattr(link, key)
+                # None != None is True, so we need to check for it specially
+                if prev != value and ~(prev is None and value is None):
```
Is this actually necessary here? I know it's a concern when we're querying SQL, but these comparisons are at the Python layer, right?

bradenmacdonald replied:
Yeah, something odd here. `None != None is True` is not accurate in python.

This part is also unchanged from master, so would it be OK if @navinkarkera opens a separate PR to clean it up? I'm not sure of the details so not confident refactoring it at the last minute.

ormsbee replied:
That's fine. But am I wrong that `has_changes` can never be False here? That seems like a straight-up bug?

bradenmacdonald replied:
Yeah I'm not sure about this code but my impression was that it's hard-coded for now until another thing is implemented elsewhere?

navinkarkera replied:
@bradenmacdonald Yes, until we store modified time in course xblocks, this was hardcoded to be True.

navinkarkera replied:
>Is this actually necessary here? I know it's a concern when we're querying SQL, but these comparisons are at the Python layer, right?
> Yeah, something odd here. None != None is True is not accurate in python.

Now, I am confused too. I remember facing with null values while comparing here but cannot remember the exact issue. I'll remove this additional check for now.

ormsbee commented on cms/djangoapps/contentstore/utils.py:
```diff
@@ -2394,3 +2394,43 @@ def create_or_update_xblock_upstream_link(xblock, course_key: str | CourseKey, c
+        try:
+            _create_or_update_container_link(course_key, created, xblock)
+        except InvalidKeyError:
+            log.error(f"Invalid key: {xblock.upstream}")
```
Nit for later follow-up: It seems bad to let this error pass with just error logging, since it likely means something more basic is broken upstream.

ormsbee commented on cms/djangoapps/contentstore/views/preview.py:
```diff
@@ -299,8 +299,14 @@ def _studio_wrap_xblock(xblock, view, frag, context, display_name_only=False):
-        can_edit = context.get('can_edit', True)
-        can_add = context.get('can_add', True)
+
+        if root_xblock and root_xblock.upstream and str(root_xblock.upstream).startswith('lct:'):
```
We should be doing comparisons on the opaque key, not substring matches.

bradenmacdonald replied:
Improved in ea131439e83ada72f5eddf439b0990069d99f0a2

ormsbee commented on cms/djangoapps/contentstore/xblock_storage_handlers/view_handlers.py:
```diff
@@ -522,6 +528,59 @@ def create_item(request):
+            notices = []
+            # Store final children keys to update order of components in unit
+            children = []
+            for i in range(len(upstream_children)):
```
Nit for later: we can use [enumerate()](https://docs.python.org/3/library/functions.html#enumerate) here.

ormsbee commented on cms/djangoapps/contentstore/xblock_storage_handlers/view_handlers.py:
```diff
@@ -522,6 +528,59 @@ def create_item(request):
+    """
+    link = UpstreamLink.get_for_block(downstream)
+    upstream_key = link.upstream_key
+    if isinstance(upstream_key, LibraryUsageLocatorV2):
+        lib_block = sync_from_upstream_block(downstream=downstream, user=request.user)
+        static_file_notices = import_static_assets_for_library_sync(downstream, lib_block, request)
+        store.update_item(downstream, request.user.id)
+    else:
```
Nit for later: Probably worth pulling the container sync into a different function and having this one call that.

bradenmacdonald replied:
I agree. Also, I want to move all the sync stuff into one `upstream_sync` app instead of having some key bits of the code randomly in `xblock_storage_handlers` (?) and the overloaded `/xblock/` REST API endpoint.

ormsbee commented on cms/djangoapps/contentstore/xblock_storage_handlers/view_handlers.py:
```diff
@@ -522,6 +528,59 @@ def create_item(request):
+                notices.append(result)
+            for child in downstream_children:
+                if child.usage_key not in children:
+                    # This downstream block was added, or deleted from upstream block.
+                    store.delete_item(child.usage_key, user_id=request.user.id)
```
Clarification: This is the part that (for the MVP) will delete any local additions to a Unit when the next upstream sync happens, right?

bradenmacdonald replied:
Yes, this deletes two different things:
1) local additions in the course - will get deleted on next sync
2) blocks deleted upstream

ormsbee commented on cms/lib/xblock/upstream_sync_block.py:
```diff
@@ -0,0 +1,174 @@
+    """
+    syncable_fields = _get_synchronizable_fields(upstream, downstream)
+    customizable_fields = set(downstream.get_customizable_fields().keys())
+    isVideoBlock = downstream.usage_key.block_type == "video"
```
Nit: I'm surprised our linting didn't complain about camel-cased var here.

bradenmacdonald replied:
Fixed. But yeah, that's suspicious....

bradenmacdonald committed:
fix: lots of "Tried to inspect a missing...upstream link" warnings

when viewing a unit in Studio

bradenmacdonald committed:
docs: mention potential REST API for future refactor

bradenmacdonald committed:
fix: check if upstream actually exists before making unit read-only

bradenmacdonald commented on cms/lib/xblock/upstream_sync.py:
```diff
@@ -96,39 +99,45 @@ def upstream_link(self) -> str | None:
-                downstream.usage_key,
-                downstream.upstream,
-            )
+            # Note: if we expect that an upstream may not be set at all (i.e. we're just inspecting a random
+            # unit that may be a regular course unit), we don't want to log this, so set log_error=False then.
+            if log_error:
```
Simply viewing a regular unit in Studio (even before this PR) was resulting in a ton of these warnings:

```
upstream_sync.py:133 - Tried to inspect an unsupported, broken, or missing downstream->upstream link: 'block-v1:BradenX+CP200+23+type@vertical+block@9d1d7e5c0b4a42958f7e4174879ec1e2'->'None'
... cms.lib.xblock.upstream_sync.NoUpstream: Content is not linked to a Content Library.
```
So I updated the code to suppress the logging in general. Most things don't have an upstream so it's expected.

bradenmacdonald committed:
chore: fix camel-case var

bradenmacdonald committed:
fix: test failure when mocked XBlock doesn't have UpstreamSyncMixin

ormsbee reviewed (dismissed):
I have a lot of nits for follow-ups, but the only blocking concern I have in the short term is the part in the models.py file where it looks like we're always saying that there are changes:

https://github.com/openedx/edx-platform/blob/4a69f30e229ad4ff44126bdada5a31ba3736fd19/cms/djangoapps/contentstore/models.py#L247-L256

@bradenmacdonald: If you've determined that this still gives us the correct behavior, then I'd really like to see some comment explaining what's going on. It's either a bug or it's really confusing in its current state.

ormsbee reviewed (approved):
I have a number of nits for follow-up, but nothing blocking.

bradenmacdonald commented:
@ormsbee 

> If you've determined that this still gives us the correct behavior, then I'd really like to see some comment explaining what's going on. It's either a bug or it's really confusing in its current state.

I didn't really get into the `ComponentLink` / link tracking personally while working on this PR. It's kind of an ancillary index just used to quickly retrieve all the library content in the course, but doesn't directly affect the core upstream sync system.

But what I believe is going on here is this function is ultimately only ever called during `XBLOCK_CREATED`/`XBLOCK_UPDATED` signal handlers (where `downstream` has changed). So there _are_ always changes of some sort when this is called.

The code is very confusing but I think what it's trying to say is this:
1. Currently, whenever you modify a downstream XBlock, it sets the `ComponentLink` "last modified at" time to the time the XBlock was modified, even if you just modified a random field like `max_attempts` which is excluded from syncing.
2. Ideally, in the future, whenever you modify a downstream XBlock, the code will change the `ComponentLink` "last modified at" time _only if_ something relevant to the upstream sync was modified, like the `upstream` source was changed, or the sync was dismissed (skip the update for this particular version), or ????????? .

bradenmacdonald commented:
Thanks everyone who collaborated on this PR, @navinkarkera @pomegranited @ormsbee !

@navinkarkera can you please work on a follow-up PR(s) for some of the cleanups that @ormsbee has identified?

navinkarkera commented:
> @ormsbee
> 
> > If you've determined that this still gives us the correct behavior, then I'd really like to see some comment explaining what's going on. It's either a bug or it's really confusing in its current state.
> 
> I didn't really get into the `ComponentLink` / link tracking personally while working on this PR. It's kind of an ancillary index just used to quickly retrieve all the library content in the course, but doesn't directly affect the core upstream sync system.
> 
> But what I believe is going on here is this function is ultimately only ever called during `XBLOCK_CREATED`/`XBLOCK_UPDATED` signal handlers (where `downstream` has changed). So there _are_ always changes of some sort when this is called.
> 
> The code is very confusing but I think what it's trying to say is this:
> 
>     1. Currently, whenever you modify a downstream XBlock, it sets the `ComponentLink` "last modified at" time to the time the XBlock was modified, even if you just modified a random field like `max_attempts` which is excluded from syncing.
> 
>     2. Ideally, in the future, whenever you modify a downstream XBlock, the code will change the `ComponentLink` "last modified at" time _only if_ something relevant to the upstream sync was modified, like the `upstream` source was changed, or the sync was dismissed (skip the update for this particular version), or ????????? .

@bradenmacdonald @ormsbee Sorry for this very confusing piece of code. You guessed it right @bradenmacdonald. At first, `has_changes` was set to False before the loop: https://github.com/openedx/edx-platform/commit/711d6aa357ddc0e888be9333b876ef08612a00ea#diff-47a27860758e70cabb21ef00cb29fb90efa277773d45546adb679ba05587f3bfR178-R185 

But while working on the frontend, I came across the requirement of sorting the links based on recently modified downstream blocks. XBlocks do not store modified dates so I updated this part to update the links modified time any time downstream block is updated and use it as base for sorting the links in frontend.

navinkarkera commented:
@bradenmacdonald @ormsbee Created https://github.com/openedx/edx-platform/pull/36599 to address comments.

ormsbee commented:
> XBlocks do not store modified dates

If you need this from the modulestore, can you call `block.edited_on` from the EditInfoMixin?

https://github.com/openedx/edx-platform/blob/e5f7a027e4ff1695454664923128ff4f4426ad91/xmodule/modulestore/edit_info.py#L22-L27

navinkarkera commented:
@ormsbee Nice! That resolves it: ac708ab5db0f

navinkarkera commented:
Created https://github.com/openedx/frontend-app-authoring/pull/1865 to compliment above change in frontend.
